src/DataStreams/CheckConstraintsBlockOutputStream.cpp

```diff
@@ -48,62 +51,56 @@ void CheckConstraintsBlockOutputStream::write(const Block & block)
 
             ColumnWithTypeAndName res_column = block_to_calculate.getByName(constraint_ptr->expr->getColumnName());
 
-            if (!isUInt8(res_column.type))
-                throw Exception(ErrorCodes::LOGICAL_ERROR, "Constraint {} does not return a value of type UInt8",
-                    backQuote(constraint_ptr->name));
+            auto result_type = removeNullable(removeLowCardinality(res_column.type));
+            auto result_column = res_column.column->convertToFullColumnIfConst()->convertToFullColumnIfLowCardinality();
 
-            if (const ColumnConst * res_const = typeid_cast<const ColumnConst *>(res_column.column.get()))
-            {
-                UInt8 value = res_const->getValue<UInt64>();
+            if (const auto * column_nullable = checkAndGetColumn<ColumnNullable>(*result_column))
+                result_column = column_nullable->getNestedColumnPtr();
```

Comment:
There is an issue: if the value of Nullable column is NULL, the value of the nested column is undetermined and we cannot rely on it. Example: `SELECT number % 2 AS n, nullIf(n, 1) AS nullable, assumeNotNull(nullable) AS unwrapped FROM numbers(10)`

Reply:
Fixed.